hadoop-hdds/docs/content/feature/ErasureCoding.md

```diff
@@ -0,0 +1,215 @@
+When client creates the file, ozone manager allocates the block group(`d + p`)
+number of nodes from the pipeline provider and return the same to client.
+As data is coming in from the application, client will write first d number of chunks
+to d number of data nodes in block group. It will also cache the d number chunks
+to generate the parity chunks. Once parity chunks generated, it will transfer the
+same to the remaining p nodes in order. Once all blocks reached their configured sizes,
+client will request the new block group nodes.
+
+Below diagram depicts the block allocation in containers as logical groups.
+For interest of space, we assumed EC(3, 2) replication config for the diagram.
+
+![EC Block Allocation in Containers](EC-Write-Block-Allocation-in-Containers.png)
+
+
+Let's zoom out the blockID: 1 data layout from the above picture, that showed in the following picture.
+This picture shows how the chunks will be layed out in data node blocks.
+![EC Chunk Layout](EC-Chunk-Layout.png)
+
+Currently ec client re-used the data transfer end-points to transfer the data to data nodes.
+That is XceiverGRPC client, used for writing data and for sending putBlock info.
+Since it used the existing transfer protocols while transferring the data, design got big advantage that,
+datanode sides changes are very minimal. The data block at data nodes would be written
+same as any other block in non-ec mode. In a single block group, container id numbers
+are same in all nodes. A file can have multiple block groups. Each block group will
+have `d+p` number of block and all ids are same.
+
+**d** - Number of data blocks in a block group
+
+**p** - Number of parity blocks in a block group
+
+### Erasure Coding Read
+
+For reads, OM will provide the node location details as part of key lookup.
+If the key is in EC, Ozone client will do the reads in EC fashion. Since the data lay out
+is different(Previous section discussed the layout), reads should consider the layout and do the reads. 
+
+EC client will open the connections to DNs based on the expected locations. When all locations are available,
+it will attempt to do plain reads chunk by chunk in round robin fashion from d data blocks.
+
+Below picture shows the order when there are no failures while reading.
+![EC Reads With no Failures](EC-Reads-With-No-Failures.png)
+
+Until it sees read failures, there is no need of doing EC reconstruction.
+
+#### Erasure Coding On-the-fly Reconstruction Reads
+
+When client detects there are failures while reading or when starting the reads,
+Ozone EC client is capable of reconstructing/recovering the lost data by doing the ec decoding.
+To do the ec decoding it needs to read parity replicas. This is a degraded read as it needs to do reconstruction.
+This reconstruction is completely transparent to the applications.
+
+Below picture depicts how it uses parity replicas in reconstruction.
+![EC Reconstructional Reads](EC-Reconstructional-Read.png)
```

Comment:
Could you change this to use the Hugo shortcode:
```
{{< image src="EC-Reconstructional-Read.png">}}
```

Reply:
Looks like many of the ozone docs followed the above approach. I already ran mvn site and it's generated fine to me.
Is there an issue with existing way to referencing? ( This is working for both intellij and site to me)
Please check this [mvn site generated file screenshot ](https://issues.apache.org/jira/secure/attachment/13039196/mvn-site-Ozone-EC-doc-screenshot.png)

Reply:
The existing approach works fine for larger screens but can cause the image to overflow out of the screen (and introduce horizontal scroll) when viewing the site on smaller screens since it does not scale down the image automatically. You can try to reproduce the behavior by going changing the viewport size by activating developer tools in chrome.

Reply:
Hi @JyotinderSingh , Thank you for providing details.
Not sure users will really view these docs in smaller screens. Dropping support to view in github and intellij is concerning me where we could allow reviewer view the images easily and check whether they embed into docs properly( without needing to run site). 
I kept all images size 5 inch range size.

If no one else has any concern, then I will go update it. Thanks for checking the docs and providing your comments.

Reply:
I think @umamaheswararao is right here, our doc uses this notation where I quickly checked, so I would suggest to stick to this notation for now.

If we want to solve the small screen problem, let's do it in a separate JIRA, I would do it differently though, as it is way more comfortable to review these files in IDE, and over Github if images are shown, so we may change the notation during site build, and than use hugo to work on the md files modified during the first step. With that we can have the advantages of both notations where we need it.

@JyotinderSingh what do you think? Would that be a feasible way to solve this problem?

Reply:
I agree with this reasoning @umamaheswararao @fapifta.
Ease of reviewing is definitely a benefit and @fapifta's suggestion of changing it during build time looks like a good solution to this problem.
I'll take this forward as another jira.

Reply:
Thanks @fapifta for chime in and providing your views.
@JyotinderSingh, very cool. Let's file separate JIRA and investigate how we could replace while building hugo. Thanks for offering the help here.